🚀 3단계 - 수강신청(DB 적용)

astrid-dm opened the pull request:
이번 리뷰도 잘 부탁 드립니다!

----
### 04.16 업데이트
피드백을 받고 적용하다보니, DM으로 언급주신 service layer의 필요성을 여로모로 깨달았습니다 🥲
이번 PR에는 service layer까지 적용하진 못했지만, 아쉬운대로 도메인 객체에 최대한 기능을 녹여냈습니다.

astrid-dm committed:
refactor : step2 코드 리뷰 반영

astrid-dm committed:
docs : readme 업데이트

astrid-dm committed:
feat : SessionRepository 추가

astrid-dm committed:
test : SessionRepositoryTest 작성

astrid-dm committed:
feat : session sql scheme 작성

astrid-dm committed:
refactor : users 객체 초기화 및 일부 property에 getter 메서드 추가

astrid-dm committed:
chore : @Deprecated 제거

seondongpyo commented on src/main/java/nextstep/courses/domain/session/FreeSession.java:
```diff
@@ -2,21 +2,29 @@
     }
 
     @Override
     public void enrollStudent(NsUser user) {
```
사용자가 이미 해당 Session에 수강신청을 했는지에 대한 여부도 검증해보면 어떨까요?

astrid-dm replied:
@seondongpyo 

``` java
    @Override
    public void enrollStudent(NsUser user) {
        if (!this.isSessionOpened()) {
            throw new IllegalArgumentException(SESSION_NOT_OPENED);
        }

        if (isPaymentAmountMatching(user)) {
            throw new IllegalArgumentException(ENROLLMENT_ALREADY_DONE);
        }

        this.users.add(user);
    }
```
무료강의의 경우, 위와 같이 수강신청을 하는 `enrollStudent` 메서드 호출시
`isPaymentAmountMatching()`로 검증을 진행해, 결제 내역이 있는 경우 이미 수강 신청이 완료된것으로 로직을 작성했습니다만,
혹시 제가 작성한 내용 외에 제가 놓친 부분이 있을까요?
아니면 말 그대로 수강신청을 기존에 했었는지를 확인하는 메서드가 별도로 존재하면 좋을지 의견 주시면 감사드리겠습니다.

seondongpyo replied:
@astrid-dm '결제 여부'가 '수강신청 여부'와 동일하다고 볼 수는 없지 않을까 생각해요.
아래처럼 동일한 사용자가 다시 수강신청을 하려는 경우도 고민해보시면 좋을 것 같습니다.

```java
@Test
@DisplayName("유료 수강 신청 가능")
void testPaidSession_isOpened() {
    PaidSession paidSession = getPaidSession(SessionStatusEnum.OPEN, 10);
    NsUser user = new NsUser();
    user.addPayment(new Payment("TEST", 1L, 1L, 10000L));

    paidSession.enrollStudent(user);
    paidSession.enrollStudent(user); // 동일한 사용자가 다시 수강신청

    assertTrue(paidSession.isSessionOpened());
    assertEquals(2, paidSession.users.getNumberOfUsers());
}
```

astrid-dm replied:
피드백 감사합니다.
`isPaymentAmountMatching` 메서드의 경우, 말 그대로 결제가 이미 완료되었는지를 확인 후 중복 결제를 막는 용도로 활용하고
별도로 (임시 메서드명) `hasEnrolled` 메서드를 생성해  이미 수강신청이 등록된 유저인지를 판단하는 로직을 따로 작성하도록 하겠습니다.

seondongpyo commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -5,31 +5,53 @@
-    protected final Users users;
+    protected int maxEnrollments = 0;
+    protected Long fee = 0L;
+    protected final Users users = new Users();
```
하나의 Session이 여러 NsUser를 가질 수도 있지만, 하나의 NsUser 또한 여러 Session을 가질 수도 있습니다.
다대다 연관관계라고 볼 수 있는데, 후자의 경우에 대한 처리도 고민해보시면 좋을 것 같아요.

astrid-dm replied:
좋은 의견 감사합니다.
다대다 연관관계를 고려하여 로직을 개선해보겠습니다.

seondongpyo commented on src/main/java/nextstep/courses/domain/session/Session.java:
```diff
@@ -5,31 +5,53 @@
     private final CoverImage coverImage;
     private final SessionStatusEnum sessionStatus;
-    protected final Users users;
+    protected int maxEnrollments = 0;
```
Nit: 인스턴스 변수는 자동으로 초기화가 되니, 꼭 명시적으로 초기화를 해줄 필요는 없을 것 같습니다.

seondongpyo reviewed (changes requested):
3단계 미션도 잘 구현해주셨네요 👍 
몇 가지 코멘트 드렸는데 확인 후 다시 리뷰 요청을 부탁 드리겠습니다 🔥

astrid-dm committed:
docs : readme 업데이트

astrid-dm committed:
feat : User 객체에 Sessions 프로퍼티 추가

Sessions = 유저가 수강한 강의 모음

astrid-dm committed:
feat : Payment 객체에 결제 금액이 동일한지 판단하는 로직 추가

astrid-dm committed:
refactor : 강의 객체의 로직 변경

- 결제 내역이 있는지 판단하는 로직은 제거 (문제에서 요구하진 않음)
- 결제 금액이 맞는지 판단하는 로직 추가 -> 이를 위해 enrollStudent()메서드의 매개변수에 Payment 추가

astrid-dm committed:
test : tc수정

seondongpyo commented on src/main/java/nextstep/courses/domain/session/FreeSession.java:
```diff
@@ -1,22 +1,50 @@
+        return !payment.isSamePriceWith(fee);
+    }
 
+    private boolean hasAlreadyEnrolled(NsUser nsUser, Session session) {
+        return nsUser.hasEnrolledSession(session);
+    }
```
메서드 내부에서 this로 바로 전달해도 될 것 같습니다.
```suggestion
    private boolean hasAlreadyEnrolled(NsUser nsUser) {
        return nsUser.hasEnrolledSession(this);
    }
```

seondongpyo reviewed (approved):
피드백 잘 반영해주셨네요.
이번 단계는 여기서 merge 하겠습니다.
다음 단계에서 뵐게요 🔥
